### PR TITLE
Implementation of #106 (TkMethods decorator)

### PR DESCRIPTION
**Pull request by cyberone**
Implementation of https://github.com/yegor256/takes/issues/106. (TkMethods)

1. Created TkMethods decorator
1. Created test case for TkMethods

### CODE REVIEW

**Comment by davvd** [comment]
@cyberone Thanks for the pull request, let me find a reviewer..

**Comment by davvd** [comment]
@carlosmiranda it's yours, please review

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    public final void canCallActOnProperMethods() throws Exception {
+        final String method = RandomStringUtils.random(10);
+        final Take take = Mockito.mock(Take.class);
+        final RqFake propreq = new RqFake(method);
+        new TkMethods(take, method).act(propreq);
+        Mockito.verify(
+            take,
+            Mockito.times(1)
+        ).act(
+                Matchers.argThat(
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Indentation here is wrong, should be aligned with line 56 above.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                    CoreMatchers.equalTo(propreq)
+                )
+            );
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Same here, indentation is wrong, too much by four spaces.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Tests if TkMethods throws HttpExcection when acting on unproper method.
+     * @throws IOException if any I/O error occurs.
+     */
+    @Test(expected = HttpException.class)
+    public final void throwsExceptionOnActinOnUnproperMethod() throws
+        IOException {
+        final String proper = RandomStringUtils.random(10);
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Can we revise this test without using random string? I don't see the need for it here.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import org.takes.rq.RqFake;
+
+/**
+ * Test case for {@link TkMethods}.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 1.0
+ * @checkstyle MagicNumberCheck (500 lines)
+ */
+public class TkMethodsTest {
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Let's make the test class `final`.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * Tests if TkMethods acts on method that is passes to it.
+     * @throws Exception if any error occurs
+     */
+    @Test
+    public final void canCallActOnProperMethods() throws Exception {
+        final String method = RandomStringUtils.random(10);
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Why is method a random string? Can we make this fixed instead?

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Test case for {@link TkMethods}.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 1.0
+ * @checkstyle MagicNumberCheck (500 lines)
+ */
+public class TkMethodsTest {
+    /**
+     * Tests if TkMethods acts on method that is passes to it.
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Please revise this Javadoc comment based on [our coding standards](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names).

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+     * @throws Exception if any error occurs
+     */
+    @Test
+    public final void canCallActOnProperMethods() throws Exception {
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Please revise this test method name based on [our coding standards](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names).

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            Mockito.times(1)
+        ).act(
+                Matchers.argThat(
+                    CoreMatchers.equalTo(propreq)
+                )
+            );
+    }
+
+    /**
+     * Tests if TkMethods throws HttpExcection when acting on unproper method.
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Should be "TkMethods can throw HttpException..." see [our coding standards here](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names).

**Comment by carlosmiranda** [comment]
@cyberone A few comments from me above.

**Comment by cyberone** [comment]
@carlosmiranda done, there's something wrong with IndentationCheck at checkstyle. I'v replaced random strings with fixed but I still think that fixed values are needed only when there's some logic depending on them.

**Comment by carlosmiranda** [comment]
@cyberone Unit tests should be as deterministic as possible. We don't want them to fail intermittently because certain values show bugs. In that case we want specifically to stick to those values so that bugs will appear consistently.

In some integration tests, we use random values because of certain constraints. But there is none such here, thus using fixed values are appropriate.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import org.takes.HttpException;
+import org.takes.Take;
+import org.takes.rq.RqFake;
+
+/**
+ * Test case for {@link TkMethods}.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 1.0
+ * @checkstyle IndentationCheck (500 lines)
```
Comment (carlosmiranda) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: @cyberone Let's move this `IndentationCheck` as close as possible to the offending code, and only use the required line count. We don't want to cover the entire file just to suppress a few lines.

**Comment by carlosmiranda** [comment]
@cyberone Looks good. Just one last comment above.

**Comment by cyberone** [comment]
@carlosmiranda done. re-check please

**Comment by carlosmiranda** [comment]
@rultor Should be good to merge

File: src/main/java/org/takes/facets/fork/TkMethods.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import org.takes.Take;
+import org.takes.tk.TkWrap;
+
+/**
+ * Take that acts on request with specified methods only.
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 1.0
```
Comment (yegor256) [inline src/main/java/org/takes/facets/fork/TkMethods.java]: it's wrong, the next version will be 0.16

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+import org.mockito.Mockito;
+import org.takes.HttpException;
+import org.takes.Take;
+import org.takes.rq.RqFake;
+
+/**
+ * Test case for {@link TkMethods}.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 1.0
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: same here

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+public final class TkMethodsTest {
+    /**
+     * TkMethods can call act on method that is passes to it.
+     * @throws Exception if any error occurs
+     */
+    @Test
+    public void callsActOnProperMethods() throws Exception {
+        final String method = "WHATEVER";
+        final Take take = Mockito.mock(Take.class);
+        final RqFake propreq = new RqFake(method);
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: why not just `req`?

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        new TkMethods(take, method).act(propreq);
+        // @checkstyle IndentationCheck (8 lines)
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: what is this for?

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+     * @throws IOException if any I/O error occurs.
+     */
+    @Test(expected = HttpException.class)
+    public void throwsExceptionOnActinOnUnproperMethod() throws
+        IOException {
+        new TkMethods(
+            Mockito.mock(Take.class),
+            "PROPER"
+        ).act(
+            new RqFake("UNPROPER")
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: why can't you put all this into the same line?

**Comment by cyberone** [comment]
@yegor256 done. There's checkstyle IndentationCheck on those lines. I don't know why, trying to reproduce, but this is single accident.

File: src/test/java/org/takes/facets/fork/TkMethodsTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+     * TkMethods can call act on method that is passes to it.
+     * @throws Exception if any error occurs
+     */
+    @Test
+    public void callsActOnProperMethods() throws Exception {
+        final String method = "WHATEVER";
+        final Take take = Mockito.mock(Take.class);
+        final RqFake req = new RqFake(method);
+        new TkMethods(take, method).act(req);
+        // @checkstyle IndentationCheck (8 lines)
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/TkMethodsTest.java]: again, why a suppression here? the code below can be formatted correctly, without this suppression

**Comment by yegor256** [comment]
@cyberone see my comment above

**Comment by cyberone** [comment]
@yegor256 removing suppression I get

    [ERROR] /src/test/java/org/takes/facets/fork/TkMethodsTest.java[55]: method call child at indentation   level 12 not at correct indentation, 16 (IndentationCheck)
    [ERROR] /src/test/java/org/takes/facets/fork/TkMethodsTest.java[56]: method call child at indentation level 16 not at correct indentation, 20 (IndentationCheck)
    [ERROR] /src/test/java/org/takes/facets/fork/TkMethodsTest.java[57]: method call rparen at indentation level 12 not at correct indentation, 16 (IndentationCheck)
    [ERROR] /src/test/java/org/takes/facets/fork/TkMethodsTest.java[58]: method call rparen at indentation level 8 not at correct indentation, 12 (IndentationCheck)

this is probably wrong.

**Comment by yegor256** [comment]
@cyberone but it doesn't mean we have to suppress qulice :) your formatting is wrong and has to be fixed, that's what qulice is telling you

**Comment by cyberone** [comment]
@yegor256 but there was complain from @carlosmiranda about indentation. There was no way to fix it without switching indentation check off.

**Comment by cyberone** [comment]
@yegor256 this is the way qulice accepts.

**Comment by cyberone** [comment]
@yegor256 should I live indentation this way or return the one with suppressed check?

**Comment by yegor256** [comment]
@cyberone do you find the code you created attractive? does it look nice and readable to you?

**Comment by cyberone** [comment]
@yegor256 done, making times(1) implicict was too hard for me for the first time.

**Comment by yegor256** [comment]
@rultor merge pls

**Comment by carlosmiranda** [comment]
@cyberone your branch has a failing test. Also, please edit the title of the PR to describe your changes as per [QA policy](www.teamed.io/qa.html).

**Comment by cyberone** [comment]
@yegor256 remerge please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by cyberone** [comment]
@yegor256 that's something but not my fault. The test fails on rultor only. Should I report a bug?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by elenavolokhova** [comment]
@davvd This one looks good.

**Comment by davvd** [comment]
> @davvd This one looks good.

@elenavolokhova thank you :)

**Comment by davvd** [comment]
@rultor deploy

**Comment by davvd** [comment]
@carlosmiranda I added 10 mins to @elenavolokhova (for QA review) in transaction 56895738... thanks, paid, **20 mins** to your account, payment ID is `56895743`, task took 144 hours and 46 mins to complete... you're getting extra minutes here (c=5)... +20 to your rating, your total score is [+4561](http://www.netbout.com/b/35023?open=rating)

**Pull request merged**